### PR TITLE
[trainer] feat: Implemented VeomniEngine as a alternative training backend

### PR DESCRIPTION
**Pull request by A1waysBeenHere**
### What does this PR do?

This PR introduces a implementation of a VeOmniEngine for VERL, providing an alternative to the existing FSDP engine.

We plan to integrate the Veomni Engine in two phases. The first phase (as part of this PR) is to complete the engine code development and conduct basic validation via SFT. The second phase is to finish the integration of the RL workflow and supplemnt the relevant document. 

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/trainer/config/_generated_ppo_trainer.yaml [hunk 1]
```diff
@@ -286,6 +286,8 @@ actor_rollout_ref:
     use_fused_kernels: false
     fused_kernel_options:
       impl_backend: torch
+    attn_implementation: flash_attention_2
```
Comment (pengwu22) [inline verl/trainer/config/_generated_ppo_trainer.yaml]: currently, for verl default hf `attn_implementation` etc. accessible by `override_config: {}` few lines above

File: verl/workers/config/model.py [hunk 2]
```diff
@@ -90,6 +90,14 @@ class HFModelConfig(BaseConfig):
...
     architectures: Optional[list[str]] = None
 
+    enable_fsdp_offload: bool = False
+    enable_reentrant: bool = False
+    attn_implementation: str = "flash_attention_2"
+    moe_implementation: str = "eager"
+    force_use_huggingface: bool = False
+    activation_gpu_limit: float = 0.0
+    basic_modules: Optional[list[str]] = field(default_factory=list)
+
```
Comment (pengwu22) [inline verl/workers/config/model.py]: some of those better be part of `VeOmniEngineConfig` instead of general HFModelConfig?

File: verl/trainer/config/optim/veomni.yaml [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# Target class for this configuration
+_target_: verl.workers.config.VeOmniOptimizerConfig
```
Comment (wuxibin89) [inline verl/trainer/config/optim/veomni.yaml:2]: Can we reuse fsdp optimizer config: verl/trainer/config/optim/fsdp.yaml?

Comment (A1waysBeenHere) [reply verl/trainer/config/optim/veomni.yaml:2]: > Can we reuse fsdp optimizer config: verl/trainer/config/optim/fsdp.yaml?

Directly use FSDP optimizer cannot work with the EP case, need to figure out why. Based on the error msg, it seems that there are 2 device_mesh when we enable EP. EP and (DP_shard, SP), they cannot share same device_mesh, and the optimizer cannot work in this suitiation. So, I might keep using VeOmni's optimizer for EP adaptation.

Comment (ji-huazhong) [reply verl/trainer/config/optim/veomni.yaml:2]: In Veomni, the parameter optimization for EP and non-EP parameters is managed separately by distinct optimizers, as seen here: https://github.com/ByteDance-Seed/VeOmni/blob/889cb3379a1143f4aa178ff55dbb3b1bcb788135/veomni/optim/optimizer.py#L311 @wuxibin89

File: verl/workers/engine/veomni/transformer_impl.py [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+                with self.model_bwd_context:
+                    loss.backward()
+
+            output_lst.append(meta_info)
+
+        return postprocess_batch_func(output_lst=output_lst, indices=indices, data=data)
+
+    def get_data_parallel_rank(self):
+        if parallel_state.get_parallel_state().ulysses_size > 1:
+            return parallel_state.get_parallel_state().device_mesh["dp"].get_local_rank()
```
Comment (vermouth1992) [inline verl/workers/engine/veomni/transformer_impl.py:254]: Is this correct with EP?

Comment (vermouth1992) [reply verl/workers/engine/veomni/transformer_impl.py:254]: Why dp_rank is torch.distributed.get_rank() when ulysses_size==1?

**Comment by vermouth1992** [comment]
Great and beautiful work!!!

File: verl/workers/engine/veomni/transformer_impl.py [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+        else:
+            return torch.distributed.get_rank()
```
Comment (vermouth1992) [inline verl/workers/engine/veomni/transformer_impl.py:256]: Why dp_rank is torch.distributed.get_rank() when ulysses_size==1?

Comment (vermouth1992) [reply verl/workers/engine/veomni/transformer_impl.py:256]: Do we support EP/PP?

Comment (A1waysBeenHere) [reply verl/workers/engine/veomni/transformer_impl.py:256]: > Do we support EP/PP?

only SP and EP works right now

**Review by wuxibin89 — APPROVED** [review]

File: verl/workers/engine/veomni/transformer_impl.py (hunk 5 above)
Comment (A1waysBeenHere) [reply verl/workers/engine/veomni/transformer_impl.py:256]: > Why dp_rank is torch.distributed.get_rank() when ulysses_size==1?

Lmao, I copyed this func with FSDPEngine style, and forgot to refine it. And yeah, I can simply `return parallel_state.get_parallel_state().device_mesh.get_local_rank("dp")`. Will fix it later
